### DISCUSSION TITLE
Unable to add self-hosted git repo as Data source

### DISCUSSION THREAD

**Discussion by sonicfather** [netbox-community/netbox#12774] [Help Wanted!]
### NetBox version

v3.5.1

### Python version

3.9

### Steps to Reproduce

1) add git repo as Data source, with credentials and https url
2) try to sync and watch logs

### Expected Behavior

Synchro of self-hosted git repo as Data source.

### Observed Behavior

Logs return an ssl error whereas custom CA certificate are installed. I try to git clone directly from the netbox worker and it is OK, only git clone using Data source does not work.

In the log we got this error:

17:14:24 default: core.jobs.sync_datasource(job=<Job: 64d84b92-962a-4f1c-8a21-0d03cc1f5670>) (64d84b92-962a-4f1c-8a21-0d03cc1f5670)
ERROR:root:Fetching remote data failed (MaxRetryError): HTTPSConnectionPool(host='mygitlab.mydomain', port=443): Max retries exceeded with url: /myproject.git/info/refs?service=git-upload-pack (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get issuer certificate (_ssl.c:1123)')))
INFO:rq.worker:default: Job OK (64d84b92-962a-4f1c-8a21-0d03cc1f5670)
17:14:24 default: Job OK (64d84b92-962a-4f1c-8a21-0d03cc1f5670)
INFO:rq.worker:Result is kept for 500 seconds

**Comment by candlerb** [comment]
The simple and secure solution would be to give your local gitlab server a valid certificate signed by LetsEncrypt. If it has a public IP (v4 or v6) and accepts inbound HTTP connections on port 80, then it's easy - use a client like certbot or dehydrated with the HTTP01 challenge.  If not, it becomes a bit more difficult as you have to set up the DNS01 challenge, either by integrating with your authoritative nameserver's API or by running a separate one like [acme-dns](https://github.com/joohoi/acme-dns).

Of course, once you have this working, you can sign all the certs you like for your domains - and automatically fetch new ones every 60-90 days.  Also consider [certgrinder](https://github.com/tykling/certgrinder) which is good if you want a single host to perform all the certificate fetching work on behalf of your other hosts; those other hosts don't need any Internet connectivity at all.

The other option would be to attempt to get your Netbox installation to trust the (private) certificate authority that you used to sign your gitlab server's certificate - which may be the server certificate itself, if it's self-signed.

For these use cases it might be nice if the "data sources" in Netbox had the option to specify a different CA certificate for each data source when using https URLs.

Since Netbox doesn't have this feature, the only other option would be to try and install the root CA certificate, but there are several places you might need to put the CA cert to get it trusted, including your system certificate store, and the python certifi package inside the Netbox virtualenv. Certificate stores vary significantly between OSes and I'm not going to guess how it might be done on yours.

**Comment by sonicfather** [comment]
Thanks for your answer.

To precise my issue, it is not a self-signed certificate but an internal PKI signed certificate. The ertificate authority  is already trusted by the netbox server.  From the server I can git clone successfully and I can also git clone from the venv.

The only "git clone" that does not worked is from the Data source feature. I was thinking about trust the certificate authority inside the certifi store but I am not confident on how to do it. Can you confirm it is the good way to resolve this issue ?

But I think by default the Data source features should trust the system certificate store.

**Reply by candlerb** [reply]
> The ertificate authority is already trusted by the netbox server.

Apparently not by Python though.  Perhaps it needs adding to the certifi store, or made available to Python some other way.

On my system I see that `/opt/netbox/venv/lib/python3.8/site-packages/certifi/cacert.pem` exists with all the root certs concatenated, and you could try adding your cert to that file as a test.

However, according to https://pypi.org/project/certifi/ :

> **Addition/Removal of Certificates**
> 
> Certifi does not support any addition/removal or other modification of the CA trust store content. This project is intended to provide a reliable and highly portable root of trust to python deployments. Look to upstream projects for methods to use alternate trust.

**Comment by sonicfather** [comment]
I added all my certificate chain (Root + sub CA) into certifi trust store, I got the same error.

**Comment by sonicfather** [comment]
Does anyone have an idea how to fix this problem?

I still do not know if I am doing somthing wrong or if it is just not possible in this netbox release.

**Comment by ChrisdAutume** [comment]
Hi There,

porcelain lib read the sslCAInfo in the gitconfig of the user to operate.

So you should use it to set your custum bundle CA:
git config http."https://xxxxx.fr/".sslCAInfo /my/path/cacert.pem

**Reply by hhopjh** [reply]
I believe this command should be executed from inside particular git repo. Correct? Where git repositories are located on Netbox internal disk?

**Comment by Wellyas** [comment]
Hi, 
I encounter the same issue with a self hosted gitlab server sign by an internal CA.
I think a way to resolve it, will be to add an other parameter for git backend, a boolean for skipping TLS verify. Base on that parameter we could add to the porcelain config dict the key http.sslVerify to match this parameter.

**Reply by candlerb** [reply]
> I think a way to resolve it, will be to add an other parameter for git backend, a boolean for skipping TLS verify.

A more secure option would be to be able to specify the CA cert - which would be the server's own cert, if it's self-signed.

e.g. for the [requests](https://requests.readthedocs.io/en/latest/user/advanced/?highlight=verify#ssl-cert-verification) library, you can either specify `verify="/path/to/cert.pem"` or `verify=False`

It would also be good to log if this *is* in fact the error.

**Reply by Wellyas** [reply]
You can do the same with [git config httpsslCAinfo](https://git-scm.com/docs/git-config#Documentation/git-config.txt-httpsslCAInfo).
The issue is how you can configure this with netbox. Because netbox use porcelain library to make git call. And each repository server have a specific git config.

in the [data_backends.py](https://github.com/netbox-community/netbox/blob/develop/netbox/core/data_backends.py) we use this method to configure a proxy.

So I think it will be nice if there is a way to add this config when we configure the data source.

If you want I can propose a PR for this.

I try to edit the data_backends to allow the TLS insecure and there is a warning log. it is not catch as standard warning but we can find it in the worker log.

Regards

**Reply by hendrikbl** [reply]
It would be nice to have the option to configure the porcelain git config when setting up the git data source.

Maintaining a custom version of `data_backends.py` will we very tedious, especially when using netbox with docker containers.

**Comment by teixemf** [comment]
I am also facing this problem and opened the feature proposal [Data Sources | Git | Allow to select customized CA Cert file for HTTPS URLs #13989](https://github.com/netbox-community/netbox/issues/13989)

**Comment by stanislav-zaprudskiy** [comment]
As under the hood `dulwich` with `urllib3` (and without `certifi`) is used, one could tell it use a "proper" CA store using `SSL_CERT_FILE` environment variable.

```
ENV SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt
```

**Reply by FloMeyer** [reply]
Hi @stanislav-zaprudskiy can you tell where to put this string?
I tried with /etc/environment with content: "SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt" But this is not working for netbox (no docker).

**Reply by markkuleinio** [reply]
My try would be to add

```
Environment="SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt"
```

inside the `[Service]` section in `/etc/systemd/system/netbox-rq.service`. 

Based on this idea: https://github.com/netbox-community/netbox/issues/15394#issuecomment-2016409907

**Reply by FloMeyer** [reply]
Thank you @markkuleinio! This is working. You made my day! :)

**Reply by ITTV-tools** [reply]
Unfortunatly this does not work for me. 
I even tried to add the file in the Dockerfile: 
```
COPY zertifikatskette.crt /usr/local/share/ca-certificates
RUN update-ca-certificates
```

An then I used the docker-override to add the Variable: 
```
  netbox-worker:
    image: netbox:latest-plugins
    environment:
      SSL_CERT_FILE: /etc/ssl/certs/ca-certificates.crt
```

Still a ssl error. 
I tried to open the page with curl inside of the container. That is working without issues

**Comment by fknorn** [comment]
So, was this ever resolved? I'm facing the same situation where my data source refuses to work because of certificate verification issues.

**Comment by netgab** [comment]
For HTTPS repos, the Python `requests` package is used, which does not use the `SSL_CERT_FILE` environment variable.
By default, `requests` is shipped with a dedicated (OS independent) trusted CA bundle. It relies on `certifi` as CA bundle source. 

You can test this, by activating your Netbox venv and start the interactive Python:
```bash
cd /opt/netbox
source venv/bin/activate
python
````

```python
>>> import requests
>>> requests.get('https://my_private_gitlab.example.test')
...
requests.exceptions.SSLError: HTTPSConnectionPool(host='my_private_gitlab.example.test', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:992)')))
```

So where is the actual CA bundle?

```python
>>> import certifi
>>> certifi.where()
'/opt/netbox-4.0.8/venv/lib/python3.11/site-packages/certifi/cacert.pem'
```

# Solution / Workaround

Do NOT manually fiddle with `/opt/netbox-4.0.8/venv/lib/python3.11/site-packages/certifi/cacert.pem` it will get overridden when the `venv` is rebuild (so on updates).

## Bare Metal installations

- Install your self-signed trusted CA certificate in your OS cert bundle (e.g. in Ubuntu / Debian `/etc/ssl/certs/ca-certificates.crt`
- Set the environment variable `REQUESTS_CA_BUNDLE` to your bundle file (e.g. `/etc/ssl/certs/ca-certificates.crt`)

The nextbox service user, must know this variable. So if you do the export with your user (`export REQUESTS_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt`).

One solution (e.g. for Debian) is to edit the file: `/etc/environment`
```bash
REQUESTS_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt
```

another way is to do a per user init script in `/etc/profile.d` (google this or ask ChatGPT :) )

## NetBox Docker

Install your self-signed trusted CA certificate in your OS cert bundle (e.g. in Ubuntu / Debian `/etc/ssl/certs/ca-certificates.crt`) of your Docker host.
Adjust your `.env` Files and `docker-compose.override.yml` as follows (shortened):

`docker-compose.override.yml`
```yml
---
services:
  netbox:
    volumes:
      - type: bind
        source: /etc/ssl/certs/ca-certificates.crt
        target: /etc/ssl/certs/ca-certificates.crt
  netbox-worker:
    volumes:
      - type: bind
        source: /etc/ssl/certs/ca-certificates.crt
        target: /etc/ssl/certs/ca-certificates.crt
```
This overrides the trusted CA certificates within the containers, with the trusted CA certificates of your Linux Docker host (which includes your private CA certificates as well).

In your NetBox environment file (`netbox.env`), add the following variable:
```bash
REQUESTS_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt
```

**Reply by grische** [reply]
@netgab thank you very much, I added this to the wiki https://github.com/netbox-community/netbox-docker/wiki/Custom-CA

**Reply by NetworkEntropologist** [reply]
I have just spent 4+ hours working on this... I have actually implemented this exact solution in my docker-compose.overide.yml file, and it made no difference at all... 

For context, my client uses a private GitLab server, which is secured using a private Enterprise CA. I have installed the full chain Root CA certificate at the host level (Ubuntu 24.04), confirmed that it is installed, checked that using a curl does not return any errors, then using this technique mapped the cert into the container and set the ENV variables correctly. However, I am still seeing the same SSL Verification Failed error.

Honestly, at this point it is becoming a time/money sink that neither I, not my client have the appetite to continue working on...

**Reply by markkuleinio** [reply]
> in my docker-compose.**overide**.yml file

Just to be sure, the file name was just a typo in the comment, right?

**Reply by NetworkEntropologist** [reply]
Oops, yes that was just a typo... NYE and all that...
